database/user_models.py

```diff
@@ -167,6 +168,52 @@ def get_private_key(self):
         from libs.s3 import get_client_private_key  # weird import triangle
         return get_client_private_key(self.patient_id, self.study.object_id)
 
+    @classmethod
+    def filtered_participants_for_study(cls, study_id: int, contains_string: str):
+        return (
+            cls.objects.filter(study_id=study_id)
+            .filter(Q(patient_id__icontains=contains_string) | Q(os_type__icontains=contains_string))
+        )
+
+    @classmethod
+    def get_values_for_view_study_table(
+            cls,
+            study_id: int,
+            start: int,
+            length: int,
+            sort_by_column_index: int,
+            sort_in_descending_order: bool,
+            contains_string: str
+    ):
+        basic_columns = ['created_on', 'patient_id', 'registered', 'os_type']
+        sort_by_column = basic_columns[sort_by_column_index]
+        if sort_in_descending_order:
+            sort_by_column = '-' + sort_by_column
+        query = (
+            cls.filtered_participants_for_study(study_id, contains_string)
+            .order_by(sort_by_column)
+            .annotate(registered=ExpressionWrapper(~Q(device_id__exact=''), output_field=BooleanField()))
+            [start: start + length])
+
+        participants_data = []
+        for participant in query:
+            # Get the list of the basic columns, which are present in every study
+            participant_values = [getattr(participant, field) for field in basic_columns]
+            # Convert the datetime object into a string in YYYY-MM-DD format
+            participant_values[0] = participant_values[0].strftime('%Y-%m-%d')
+            # Add all values for intervention dates
+            for intervention_date in participant.intervention_dates.order_by(Lower('intervention__name')):
+                if intervention_date.date is not None:
+                    participant_values.append(intervention_date.date.strftime('%Y-%m-%d'))
+                else:
+                    participant_values.append(None)
+            # Add all values for custom fields
+            for custom_field_val in participant.field_values.order_by(Lower('field__field_name')):
```

Comment:
:information_source: Given your testing, not necessary to do, but running `.intervention_dates` or `order_by('interviention__name')` or `order_by('field__field_name')` will make a separate request for each participant. This can be avoided if you have more performance problems using `select_related` or `prefetch_related` (https://stackoverflow.com/a/31237071/1940450).

Reply:
Done using `prefetch_related`.  Thanks for the suggestion.  It took me a little while to figure out how to sort on a related field inside the `prefetch_related` call, but I got it done.  I checked, and it did cut the number of database calls.  This AJAX query now executes in a constant of 6 database calls, regardless of the number of participants it pulls.